Archlinux build fails

**Describe the bug**
CMake's configure stage fails with:
```
-- Using prebuilt third-party library Filament
-- GLIBC version 2.33 found: Downloading Open3D Filament binary.
-- Filament is located at /home/zannzetsu/Open3D/build/filament/src/ext_filament
CMake Error at 3rdparty/find_dependencies.cmake:971 (find_library):
  Could not find CPPABI_LIBRARY using the following names: c++abi
Call Stack (most recent call first):
  CMakeLists.txt:613 (include)
```
This applies to both 0.12 and master.

**To Reproduce**
```
git clone --recurse-submodules --depth=1 git@github.com:intel-isl/Open3D.git
cd Open3D
mkdir build
cd build
cmake ..
```
**Expected behavior**
Configure succeeds along with the following make.

**Environment (please complete the following information):**

- Operating system: Arch Linux
- Python version: Python 3.9.5
- Open3D version: Currently installed is 0.9 but I wanted to install 0.12 or master
- Is this remote workstation?: no
- How did you install Open3D?: 0.9 from AUR, 0.12 from source
- Compiler version (if built from source): gcc (GCC) 11.1.0

**Additional information**
find_dependencies fails here
```
        if (NOT CLANG_LIBDIR)
            find_library(CPPABI_LIBRARY c++abi PATH_SUFFIXES
                         llvm-11/lib llvm-10/lib llvm-9/lib llvm-8/lib llvm-7/lib
                         REQUIRED)
```
Here is the output of `llvm-as --version`
```
LLVM (http://llvm.org/):
  LLVM version 11.1.0
  Optimized build.
  Default target: x86_64-pc-linux-gnu
  Host CPU: skylake
  ```
 I can install `pacman -S libc++abi` but the version is `libc++abi 12.0.0-1`

I just tried and the error message is:
```
-- Using prebuilt third-party library Filament
-- GLIBC version 2.33 found: Downloading Open3D Filament binary.
-- Filament is located at /home/zannzetsu/Open3D/build/filament/src/ext_filament
CMake Error at 3rdparty/find_dependencies.cmake:977 (find_library):
  Could not find CPP_LIBRARY using the following names: c++
Call Stack (most recent call first):
  CMakeLists.txt:613 (include)
```
I'm not quite sure where to go from here, no obvious sources of this error come to my mind.

You are right, installing `libc++` did allow cmake to configure the project. Now I am facing compile issues.
```
[ 29%] Building CXX object cpp/open3d/core/CMakeFiles/core.dir/nns/NanoFlannIndex.cpp.o                                                                                          │
In file included from /home/zannzetsu/Open3D/3rdparty/parallelstl/include/pstl/execution:20,                                                                                     │
                 from /home/zannzetsu/Open3D/cpp/open3d/utility/ParallelScan.h:32,                                                                                               │
                 from /home/zannzetsu/Open3D/cpp/open3d/core/nns/NanoFlannIndex.cpp:36:                                                                                          │
/home/zannzetsu/Open3D/3rdparty/parallelstl/include/pstl/internal/execution_defs.h:29:7: error: redefinition of ‘class __pstl::execution::v1::sequenced_policy’                  │
   29 | class sequenced_policy                                                                                                                                                   │
      |       ^~~~~~~~~~~~~~~~                                                                                                                                                   │
In file included from /usr/include/c++/11.1.0/pstl/glue_memory_defs.h:13,                                                                                                        │
                 from /usr/include/c++/11.1.0/memory:144,                                                                                                                        │
                 from /home/zannzetsu/Open3D/cpp/open3d/core/Tensor.h:31,                                                                                                        │
                 from /home/zannzetsu/Open3D/cpp/open3d/core/nns/NanoFlannIndex.h:31,                                                                                            │
                 from /home/zannzetsu/Open3D/cpp/open3d/core/nns/NanoFlannIndex.cpp:27:                                                                                          │
/usr/include/c++/11.1.0/pstl/execution_defs.h:23:7: note: previous definition of ‘class __pstl::execution::v1::sequenced_policy’                                                 │
   23 | class sequenced_policy                                                                                                                                                   │
      |       ^~~~~~~~~~~~~~~~                                                                                                                                                   │
In file included from /home/zannzetsu/Open3D/3rdparty/parallelstl/include/pstl/execution:20,                                                                                     │
                 from /home/zannzetsu/Open3D/cpp/open3d/utility/ParallelScan.h:32,                                                                                               │
                 from /home/zannzetsu/Open3D/cpp/open3d/core/nns/NanoFlannIndex.cpp:36:                                                                                          │
/home/zannzetsu/Open3D/3rdparty/parallelstl/include/pstl/internal/execution_defs.h:52:7: error: redefinition of ‘class __pstl::execution::v1::parallel_policy’                   │
   52 | class parallel_policy                                                                                                                                                    │
      |       ^~~~~~~~~~~~~~~ 
```
and multiple similar errors after that. Should I open a seperate issue?

I tried building with clang but got a different error instead:
```
[ 74%] Built target Open3D
[ 74%] Building CXX object cpp/tools/CMakeFiles/ViewGeometry.dir/ViewGeometry.cpp.o
make[2]: *** Brak reguł do zrobienia obiektu 'libjpeg-turbo-install/lib/libturbojpeg.a', wymaganego przez 'bin/ViewGeometry'. Stop.
make[1]: *** [CMakeFiles/Makefile2:2171: cpp/tools/CMakeFiles/ViewGeometry.dir/all] Błąd 2
make: *** [Makefile:136: all] Błąd 2
```

This translates roughly to "No rule to create object 'libjpeg-turbo-install/lib/libturbojpeg.a' required by 'bin/ViewGeometry'.

I do have libturbojpeg installed.
`sudo pacman -Q | grep libjpeg`:
```
lib32-libjpeg-turbo 2.1.0-1
libjpeg-turbo 2.1.0-1
```
I appreciate the work you put into solving this problem and understand that if this is an arch centric issue I will have to solve it myself.

New error message:
```
[ 27%] Built target core
Consolidate compiler generated dependencies of target geometry
[ 27%] Building CXX object cpp/open3d/geometry/CMakeFiles/geometry.dir/Line3D.cpp.o
/home/zannzetsu/Open3D/cpp/open3d/geometry/Line3D.cpp:30:11: error: no member named 'transform' in 'open3d::geometry::Line3D'
    this->transform(t);
    ~~~~  ^
/home/zannzetsu/Open3D/cpp/open3d/geometry/Line3D.cpp:324:11: error: no member named 'transform' in 'open3d::geometry::Segment3D'
    this->transform(t);
    ~~~~  ^
2 errors generated.
make[2]: *** [cpp/open3d/geometry/CMakeFiles/geometry.dir/build.make:202: cpp/open3d/geometry/CMakeFiles/geometry.dir/Line3D.cpp.o] Błąd 1
make[1]: *** [CMakeFiles/Makefile2:1298: cpp/open3d/geometry/CMakeFiles/geometry.dir/all] Błąd 2
make: *** [Makefile:136: all] Błąd 2

```

--- a/3rdparty/find_dependencies.cmake
+++ b/3rdparty/find_dependencies.cmake
@@ -917,8 +917,8 @@ if(BUILD_GUI)
             endif()
             # If the default version is not sufficient, look for some specific versions
             if(NOT FILAMENT_C_COMPILER OR NOT FILAMENT_CXX_COMPILER)
-                find_program(CLANG_VERSIONED_CC NAMES clang-11 clang-10 clang-9 clang-8 clang-7)
-                find_program(CLANG_VERSIONED_CXX NAMES clang++11 clang++-10 clang++-9 clang++-8 clang++-7)
+                find_program(CLANG_VERSIONED_CC NAMES clang-12 clang-11 clang-10 clang-9 clang-8 clang-7)
+                find_program(CLANG_VERSIONED_CXX NAMES clang++-12 clang++11 clang++-10 clang++-9 clang++-8 clang++-7)
                 if (CLANG_VERSIONED_CC AND CLANG_VERSIONED_CXX)
                     set(FILAMENT_C_COMPILER "${CLANG_VERSIONED_CC}")
                     set(FILAMENT_CXX_COMPILER "${CLANG_VERSIONED_CXX}")
@@ -969,19 +969,13 @@ if(BUILD_GUI)
         # Find CLANG_LIBDIR if it is not defined. Mutiple paths will be searched.
         if (NOT CLANG_LIBDIR)
             find_library(CPPABI_LIBRARY c++abi PATH_SUFFIXES
-                         llvm-11/lib llvm-10/lib llvm-9/lib llvm-8/lib llvm-7/lib
+                         llvm-12/lib llvm-11/lib llvm-10/lib llvm-9/lib llvm-8/lib llvm-7/lib
                          REQUIRED)
             get_filename_component(CLANG_LIBDIR ${CPPABI_LIBRARY} DIRECTORY)
         endif()
         # Find clang libraries at the exact path ${CLANG_LIBDIR}.
         find_library(CPP_LIBRARY    c++    PATHS ${CLANG_LIBDIR} REQUIRED NO_DEFAULT_PATH)
         find_library(CPPABI_LIBRARY c++abi PATHS ${CLANG_LIBDIR} REQUIRED NO_DEFAULT_PATH)
-        if(CPP_LIBRARY-NOTFOUND)
-            message(FATAL_ERROR "CPP_LIBRARY-NOTFOUND")
-        endif()
-        if(CPPABI_LIBRARY-NOTFOUND)
-            message(FATAL_ERROR "CPPABI_LIBRARY-NOTFOUND")
-        endif()
         # Ensure that libstdc++ gets linked first
         target_link_libraries(3rdparty_filament INTERFACE -lstdc++
                               ${CPP_LIBRARY} ${CPPABI_LIBRARY})

--- a/cpp/open3d/utility/ParallelScan.h
+++ b/cpp/open3d/utility/ParallelScan.h
@@ -28,9 +28,28 @@
 
 #include <tbb/parallel_for.h>
 #include <tbb/parallel_scan.h>
+
 #if TBB_INTERFACE_VERSION >= 10000
+
+// Check if the C++ standard library implements parallel algorithms
+// and use this over parallelstl to avoid conflicts.
+// Clang does not implement it so far, so checking for C++17 is not sufficient.
+#ifdef __cpp_lib_parallel_algorithm
+#include <execution>
+#include <numeric>
+#else
 #include <pstl/execution>
 #include <pstl/numeric>
+
+// parallelstl incorrectly assumes MSVC to unconditionally implement
+// parallel algorithms even if __cpp_lib_parallel_algorithm is not defined.
+// So manually include the header which pulls all "pstl::execution" definitions
+// into the "std" namespace.
+#if __PSTL_CPP17_EXECUTION_POLICIES_PRESENT
+#include <pstl/internal/glue_execution_defs.h>
+#endif
+
+#endif
 #endif
 
 namespace open3d {
@@ -66,7 +85,7 @@ template <class Tin, class Tout>
 void InclusivePrefixSum(const Tin* first, const Tin* last, Tout* out) {
 #if TBB_INTERFACE_VERSION >= 10000
     // use parallelstl if we have TBB 2018 or later
-    std::inclusive_scan(pstl::execution::par_unseq, first, last, out);
+    std::inclusive_scan(std::execution::par_unseq, first, last, out);
 #else
     ScanSumBody<Tin, Tout> body(out, first);
     size_t n = std::distance(first, last);

